### PR TITLE
Porting to STM32F303 and STM32F3Discovery kit

### PR DESCRIPTION
**Pull request by alexandruradovici**
### Pull Request Overview

This pull request adds initial support for the stm32f3xx chip and the stm32f3discovery board.
It is useful as the stm32f3discovery boards are very useful for use in class being inexpensive and having most of the features required for mcu labs and trainings.

Features:
  * Button
  * LEDs
  * USART1 (console writing and reading)
  * GPIO
  * EXTI
  * alarm (for delay)
  * SPI
  * L3GD20 (waiting for https://github.com/tock/tock/pull/1723 to be merged first)
  * I2C almost done (waiting for I2C component https://github.com/tock/tock/pull/1725 to be merged)

Still in progress: usart via dma, adc and other peripherals.

This was inspired from the stm32f4xx chip and stm boards.

Any feedback would be greatly appreciated.

### Testing Strategy

This pull request was tested with an stm32f3 discovery kit.

### Documentation Updated

- [x] Updated the relevant files in `/docs`.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: chips/stm32f3xx/src/deferred_call_tasks.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+//! Deferred calls allow peripheral drivers to register pseudo interrupts.
+//! These are the definitions of which deferred calls this chip needs.
+
+use core::convert::Into;
+use core::convert::TryFrom;
+
+/// A type of task to defer a call for
+#[derive(Copy, Clone)]
+pub enum Task {
+    Nop = 0,
```
Comment (bradjc) [inline chips/stm32f3xx/src/deferred_call_tasks.rs]: What is a `Nop` task?

File: boards/stm32f3discovery/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+            (
+                stm32f3xx::gpio::PinId::PE13.get_pin().as_ref().unwrap(),
+                capsules::led::ActivationMode::ActiveHigh
+            ),
+        ]
+    );
+    let led = static_init!(
+        capsules::led::LED<'static>,
+        capsules::led::LED::new(&led_pins[..])
+    );
```
Comment (bradjc) [inline boards/stm32f3discovery/src/main.rs]: Please update to the component.

File: boards/stm32f3discovery/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+            kernel::hil::gpio::FloatingState::PullNone
+        )),
+    );
+
+    // ALARM
+    let mux_alarm = static_init!(
+        MuxAlarm<'static, stm32f3xx::tim2::Tim2>,
+        MuxAlarm::new(&stm32f3xx::tim2::TIM2)
+    );
+    stm32f3xx::tim2::TIM2.set_client(mux_alarm);
```
Comment (bradjc) [inline boards/stm32f3discovery/src/main.rs]: Please update to the component.

File: chips/stm32f3xx/src/usart.rs [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+        self.enable_transmit_complete_interrupt();
+    }
+}
+
+impl hil::uart::Transmit<'a> for Usart<'a> {
+    fn set_transmit_client(&self, client: &'a dyn hil::uart::TransmitClient) {
+        self.tx_client.set(client);
+    }
+
+    // TODO
```
Comment (bradjc) [inline chips/stm32f3xx/src/usart.rs]: Need implementation

File: chips/stm32f3xx/src/deferred_call_tasks.rs (hunk 1 above)
Comment (alexandruradovici) [reply chips/stm32f3xx/src/deferred_call_tasks.rs]: I based my implementation on the stm32f4xx, this was there. Will look into it, not sure at what it was used.

File: boards/stm32f3discovery/src/main.rs (hunk 2 above)
Comment (alexandruradovici) [reply boards/stm32f3discovery/src/main.rs]: I'm not sure I understand, this is how stm32f4xx is done. Probably I am missing something.

File: boards/stm32f3discovery/src/main.rs (hunk 3 above)
Comment (alexandruradovici) [reply boards/stm32f3discovery/src/main.rs]: Same as previous.

File: boards/stm32f3discovery/src/main.rs (hunk 2 above)
Comment (alevy) [reply boards/stm32f3discovery/src/main.rs]: It looks like the other stm32 boards have not been fully updated to use components (in praticular, the LED driver and alarm haven't but some others have been).

You can look at other boards (Hail and imix) for how this _should_ be done, and it's maybe worth updating the nucleo boards in a separate PR.

Comment (alexandruradovici) [reply boards/stm32f3discovery/src/main.rs]: Got it already, will fix for discovery and nucleo boards.

File: chips/stm32f3xx/src/deferred_call_tasks.rs (hunk 1 above)
Comment (alexandruradovici) [reply chips/stm32f3xx/src/deferred_call_tasks.rs]: I think this is just an example of a task, should I remove it?

File: chips/stm32f3xx/src/usart.rs (hunk 4 above)
Comment (alexandruradovici) [reply chips/stm32f3xx/src/usart.rs]: I stil need some time to figure out how DMA works on F3, will do asap.

File: boards/stm32f3discovery/src/main.rs (hunk 2 above)
Comment (alexandruradovici) [reply boards/stm32f3discovery/src/main.rs]: Done

File: boards/stm32f3discovery/src/main.rs (hunk 3 above)
Comment (alexandruradovici) [reply boards/stm32f3discovery/src/main.rs]: Done###

File: boards/stm32f3discovery/src/io.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            &mut [led],
+            writer,
+            info,
+            &cortexm4::support::nop,
+            &PROCESSES,
+            &CHIP,
+        )
+    });
+
+    loop {}
```
Comment (alevy) [inline boards/stm32f3discovery/src/io.rs]: This loop shouldn't necessary. `debug::panic` is already `-> !` and never returns.

File: boards/stm32f3discovery/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+    let button = components::button::ButtonComponent::new(board_kernel).finalize(
+        components::button_component_helper!((
+            stm32f3xx::gpio::PinId::PA00.get_pin().as_ref().unwrap(),
+            kernel::hil::gpio::ActivationMode::ActiveLow,
+            kernel::hil::gpio::FloatingState::PullNone
+        )),
+    );
+
+    // ALARM
+    // let mux_alarm = static_init!(
```
Comment (alevy) [inline boards/stm32f3discovery/src/main.rs]: Remove old (commented) code

File: boards/stm32f3discovery/src/io.rs (hunk 5 above)
Comment (alexandruradovici) [reply boards/stm32f3discovery/src/io.rs]: It seems that stm32f4 has the same issue, should I fixed that also?

File: chips/stm32f3xx/src/usart.rs (hunk 4 above)
Comment (alexandruradovici) [reply chips/stm32f3xx/src/usart.rs]: Done, for now without DMA, using interrupts.

**Comment by alexandruradovici** [comment]
I added EXT support, I think the PR meets the minimum required for merging.

**Comment by alexandruradovici** [comment]
Any feedback here? Is there something else missing that is preventing the merge? Thanks.

**Comment by alexandruradovici** [comment]
Fixed the typos, sorry for that.

**Comment by alexandruradovici** [comment]
Added SPI functionality, is there any way you could merge this? Thank you.

File: boards/stm32f3discovery/Makefile [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+OPENOCD_OPTIONS=-f openocd.cfg
+
+# OpenOCD requires fullpath
+CWD=$(shell pwd)
+
+# sudo is needed for libusb access to work properly
+
+.PHONY: flash-debug
+flash-debug: target/$(TARGET)/debug/$(PLATFORM).elf
+	sudo $(OPENOCD) $(OPENOCD_OPTIONS) -c "init; reset halt; flash write_image erase $(CWD)/$<; verify_image $(CWD)/$<; reset; shutdown"
```
Comment (bradjc) [inline boards/stm32f3discovery/Makefile]: This makes me nervous. I think we should remove the sudo from the flash commands, and point people to https://elinux.org/Accessing_Devices_without_Sudo instead.

Comment (alexandruradovici) [reply boards/stm32f3discovery/Makefile]: Sorry for that, I based my implementation on the nucleo boards, they both have it. I removed the sudo.

File: boards/stm32f3discovery/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,429 @@
+//! Board file for Nucleo-F446RE development board
+//!
+//! - <https://www.st.com/en/evaluation-tools/nucleo-f446re.html>
```
Comment (bradjc) [inline boards/stm32f3discovery/src/main.rs]: Update

File: boards/stm32f3discovery/src/io.rs [hunk 9]
```diff
@@ -0,0 +1,79 @@
+use core::fmt::Write;
+use core::panic::PanicInfo;
+
+// use cortexm4;
```
Comment (ppannuto) [inline boards/stm32f3discovery/src/io.rs:4]:
```suggestion
```

File: chips/stm32f3xx/src/deferred_call_tasks.rs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+//! Deferred calls allow peripheral drivers to register pseudo interrupts.
+//! These are the definitions of which deferred calls this chip needs.
+
+use core::convert::Into;
+use core::convert::TryFrom;
+
+/// A type of task to defer a call for
+#[derive(Copy, Clone)]
+pub enum Task {
+    // ... add deffred tasks here
```
Comment (ppannuto) [inline chips/stm32f3xx/src/deferred_call_tasks.rs:12]:
```suggestion
    // ... add deferred tasks here
```

File: boards/README.md [hunk 11]
```diff
@@ -4,16 +4,17 @@ Platforms Supported by Tock
...
+| [Hail](hail/README.md)                               | ARM Cortex-M4   | SAM4LC8BA      | Bootloader | tockloader     |
+| [Imix](imix/README.md)                               | ARM Cortex-M4   | SAM4LC8CA      | Bootloader | tockloader     |
+| [Nordic nRF52-DK](nordic/nrf52dk/README.md)          | ARM Cortex-M4   | nRF52832       | jLink      | tockloader     |
+| [Nordic nRF52840-DK](nordic/nrf52840dk/README.md)    | ARM Cortex-M4   | nRF52840       | jLink      | tockloader     |
+| [ACD52832](acd52832/README.md)                       | ARM Cortex-M4   | nRF52832       | jLink      | tockloader     |
+| [TI LAUNCHXL-CC26x2](launchxl/README.md)             | ARM Cortex-M4   | CC2652R        | openocd    | tockloader     |
+| [ST Nucleo F446RE](nucleo_f446re/README.md)          | ARM Cortex-M4   | STM32F446      | openocd    | custom         |
+| [ST Nucleo F429ZI](nucleo_f429zi/README.md)          | ARM Cortex-M4   | STM32F429      | openocd    | custom         |
+| [STM32F3Discovery kit](stm32f3discovery/README.md)   | ARM Cortex-M3   | STM32F303VCT6  | openocd    | custom         |
```
Comment (brghena) [inline boards/README.md:17]:
```suggestion
| [STM32F3Discovery kit](stm32f3discovery/README.md)   | ARM Cortex-M4   | STM32F303VCT6  | openocd    | custom         |
```

Comment (brghena) [reply boards/README.md:17]: Unless I'm very mistaken, the STM32F3 Discovery kit is a Cortex-M4 processor.

Comment (alexandruradovici) [reply boards/README.md:17]: Yes, my mistake, sorry.

File: boards/stm32f3discovery/openocd.cfg [hunk 12]
```diff
@@ -0,0 +1,18 @@
+#interface
+interface hla
+hla_layout stlink
+hla_device_desc "ST-LINK/V2-1"
+hla_vid_pid 0x0483 0x374b
+
+# The chip has 40KB sram
```
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply boards/stm32f3discovery/openocd.cfg:7]: Yes, thanks for the feedback.

**Comment by lschuermann** [comment]
> I know the other stm32 has them, but is the feature necessary? Does it really matter if a few more entries are defined that won't be used on a particular chip?

@bradjc I don't find an answer to that question. You're talking about the deferred calls, right? In my humble opinion if the chip doesn't need them that should be removed and re-added if required. Most things outside of the chip crate should be able to use the dynamic deferred calls anyways. Also, by referring to the `Task` enum's `Example` variant, we're shipping dead and (deliberately) broken code.

**Pull request merged**